### ISSUE TITLE
Publish policy on using generics in gateway-api

### ISSUE THREAD

**Issue by keithmattix** [kubernetes-sigs/gateway-api#1388]
**What would you like to be added**:
With generics shipping in go1.18, the gateway-api project needs to have some kind of policy around when and how generics should be used. If upstream Kubernetes has a policy, we likely need to mirror it. If further discussion is needed, please use GitHub Discussions and/or Slack to engage members of the community.

**Why this is needed**:
Generics are a very new Go feature, and the implications of using them in a large codebase/project are unknown. Thus, it will be useful for contributors to have guidelines for how to use generics in the gateway-api codebase.

/good-first-issue

**Comment by akankshakumari393** [comment]
I would like to work on this issue. But would need guidance as well
/assign

**Comment by mikemorris** [comment]
Refs https://github.com/kubernetes/community/blob/master/sig-architecture/generics.md for the upstream Kubernetes policy, and https://github.com/kubernetes/kubernetes/issues/106846 as a specific attempt at introducing generics in a meaningful way.

FWIW I did previously introduce some trivial usage of generics in https://github.com/kubernetes-sigs/gateway-api/pull/1081

**Comment by akankshakumari393** [comment]
We have a upstream Kubernetes policy in place for using generics, I think we should mirror it for single source of truth

**Comment by shaneutt** [comment]
Where are we at with this?

**Comment by shaneutt** [comment]
@EyalPazz so I wanna say yes, but reviewing this I think there's a bit of nuance and I'm not really sure it's so definitively a `good-first-issue`. If you wanna take this, we'll need to talk it over in a Slack thread or on one of our regular meetings to kinda "revive it" since it's a few years old now and I think we need to sort of discuss what value we want to get out of this, I wouldn't say it's quite "ready for work" at least not without a bit of chatting it over 🤔

**Comment by shaneutt** [comment]
@EyalPazz yes, but I think the next best step here wouldn't be to go straight into implementation but maybe since it's gotten quite old we should talk this over a bit either in our Slack channel or (preferably) as an agenda item for one of our calls, just to get a bit more information and make sure we have our bearings. 🤔

**Comment by mikemorris** [comment]
Discussion in an upcoming meeting seems good to me - I think this should be much less controversial to move to implementation now that we're several Golang releases since generics were introduced.
